blocks/bricks/__init__.py

```diff
@@ -494,45 +495,73 @@ def apply(self, input_):
         return tensor.nnet.softmax(input_)
 
-    @application
-    def categorical_cross_entropy(self, y, x):
-        """Return computationally stable softmax cost.
+    @application(inputs=['input_'], outputs=['output'])
+    def log_probabilities(self, input_):
+        """Normalize log-probabilities.
+
+        Converts unnormalized log-probabilities (exponents of which do not
+        sum to one) into actual log-probabilities (exponents of which sum
+        to one).
+
+        Parameters
+        ----------
+        input_ : :class:`~theano.Variable`
+            A matrix, each row contains unnormalized log-probabilities of a
+            distribution.
+
+        Returns
+        -------
+        output : :class:`~theano.Variable`
+            A matrix with normalized log-probabilities for each
+            distribution from `input_`.
+
+        """
+        shifted = input_ - input_.max(axis=1, keepdims=True)
+        return shifted - tensor.log(
+            tensor.exp(shifted).sum(axis=1, keepdims=True))
+
+    @application(inputs=['y', 'x'], outputs=['output'])
+    def categorical_cross_entropy(self, application_call, y, x):
+        """Computationally stable cross-entropy for pre-softmax values.
 
         Parameters
         ----------
         y : :class:`~tensor.TensorVariable`
             In the case of a matrix argument, each slice along
-            axis represents one distribution. In the vector case, each
+            axis 0 represents one distribution. In the vector case, each
             element represents the position of the '1' in a one hot-vector.
         x : :class:`~tensor.TensorVariable`
-            Each slice along axis represents energies of a distribution,
-            that is pre-softmax values.
+            Each slice along axis 0 represents unnormalized
+            log-probabilities of a distribution, that is pre-softmax
+            values.
 
         Returns
         -------
         cost : :class:`~tensor.TensorVariable`
-            The cross entropy between y and x.
+            A vector of cross-entropies between distributions y and x.
 
         """
-        x = x - x.max(axis=1).dimshuffle(0, 'x')
-        log_prob = x - tensor.log(tensor.exp(x).sum(axis=1).dimshuffle(0, 'x'))
+        x = self.log_probabilities(x)
+        application_call.add_auxiliary_variable(
+            named_copy(x, 'log_probabilities'))
         if y.ndim == x.ndim - 1:
-            flat_log_prob = log_prob.flatten()
-            range_ = tensor.arange(y.shape[0])
-            flat_indices = y + range_ * x.shape[1]
-            cost = -tensor.mean(flat_log_prob[flat_indices])
+            cost = -x[tensor.arange(y.shape[0]), y]
         elif y.ndim == x.ndim:
-            cost = -tensor.mean((log_prob * y).sum(axis=1))
+            cost = -(x * y).sum(axis=1)
         else:
             raise TypeError('rank mismatch between x and y')
         return cost
 
 
+class SoftmaxWithExtraDims(Softmax):
```

Comment:
I like your `NDimensionalSoftmax`. Mainly because this is not the end of the story, and we need more capabilities for this brick (e.g. to softmax along a custom axis). I believe the exact formula might change, that is another wrapper or a few other wrappers will be used to create `NDimensionalSoftmax` from `Softmax`. 

In fact we could concentrate lots of options for reshaping and reshuffling in one wrapper, that might be easier than `WithExtraDims`, `WithAxisSwapped`, `AlongSpecifiedAxis` and etc.